File: windows-winui/YtdlWin.Tests/NoticeTests.cs
/* What the queue announces, and when -- Core/Notices.cs's rules as fixtures.
 *
 * THE SAME VALUES are asserted by linux-gtk/tests/test_notify.c and
 * macos-swiftui/Tests/NoticeTests.swift. A title or a rule changed here and
 * not there is three apps announcing the same queue three different ways, so
 * the strings are spelled out in full rather than matched loosely.
 *
 * The last test drives the real Runner, to pin the one thing the tracker
 * cannot check for itself: that Runner.Settled never reports nothing left to
 * run while a run is about to start.
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using Xunit;
using YtdlWin.Core;

namespace YtdlWin.Tests;

public sealed class NoticeTests
{
    private static RunRecord Rec(string id, string state, string lastLine = "",
                                 long touched = -1, long skipped = -1, long errors = -1)
    {
        var url = $"https://youtu.be/{id}";
        return new RunRecord
        {
            Id = id, State = state, Command = $"ytdl {url}", LastLine = lastLine,
            VideosTouched = touched, ArchiveSkipped = skipped, Errors = errors, Warnings = 0,
            ExitCode = state == "failed" ? 1 : 0,
            Opts = new RunOptions { Url = url },
        };
    }

    [Fact]
    public void SingleDone()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        Assert.Equal(new Notice("Download finished",
                                "https://youtu.be/aaaaaaaaaaa\n3 touched, 12 already archived",
                                false),
                     t.Update(new[] { Rec("aaaaaaaaaaa", "done", touched: 3, skipped: 12, errors: 0) }, 0));
    }

    [Fact]
    public void SingleFailed()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        Assert.Equal(new Notice("Download failed",
                                "https://youtu.be/bbbbbbbbbbb\n"
                                + "ERROR: [youtube] bbbbbbbbbbb: Private video",
                                true),
                     t.Update(new[] { Rec("bbbbbbbbbbb", "failed",
                                          "ERROR: [youtube] bbbbbbbbbbb: Private video") }, 0));
    }

    /// Three runs, one at a time, the middle one failing: silence, the failure
    /// at once, then one summary that repeats it.
    [Fact]
    public void QueueOfThree()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        var c1 = Rec("c1", "done", touched: 1, skipped: 0, errors: 0);
        var c2 = Rec("c2", "failed", "ERROR: HTTP Error 403: Forbidden", touched: 0, skipped: 0, errors: 1);
        var c3 = Rec("c3", "done", touched: 2, skipped: 5, errors: 0);

        Assert.Null(t.Update(new[] { c1 }, 2));
        Assert.Equal(new Notice("A download failed",
                                "https://youtu.be/c2\nERROR: HTTP Error 403: Forbidden\n"
                                + "The queue goes on: 1 still to run.",
                                true),
                     t.Update(new[] { c2, c1 }, 1));
        // Nothing new: nothing to say, and the failure is not announced twice.
        Assert.Null(t.Update(new[] { c2, c1 }, 1));
        Assert.Equal(new Notice("Queue finished: 2 done, 1 failed",
                                "3 touched, 5 already archived, 1 error\n"
                                + "Last failure: https://youtu.be/c2 — "
                                + "ERROR: HTTP Error 403: Forbidden",
                                true),
                     t.Update(new[] { c3, c2, c1 }, 0));
    }

    [Fact]
    public void SecondFailureCounts()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        var f1 = Rec("f1", "failed", "ERROR: one");
        var f2 = Rec("f2", "failed", "ERROR: two");
        Assert.Equal("A download failed", t.Update(new[] { f1 }, 3)?.Title);
        Assert.Equal(new Notice("2 downloads failed so far",
                                "https://youtu.be/f2\nERROR: two\nThe queue goes on: 2 still to run.",
                                true),
                     t.Update(new[] { f2, f1 }, 2));
    }

    /// A run finishing and the queue going idle between two polls: the
    /// summary only, never a failure toast followed by a summary.
    [Fact]
    public void CoalescedTick()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        Assert.Equal(new Notice("Queue finished: 1 done, 1 failed",
                                "4 touched, 0 already archived\n"
                                + "Last failure: https://youtu.be/d2 — ERROR: gone",
                                true),
                     t.Update(new[] { Rec("d2", "failed", "ERROR: gone"),
                                      Rec("d1", "done", touched: 4, skipped: 0, errors: 0) }, 0));
    }

    [Fact]
    public void CancelledOnlyIsSilent()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        var e1 = Rec("e1", "cancelled");
        Assert.Null(t.Update(new[] { e1 }, 0));
        // ...and the queue it ended is closed: the next one does not count it.
        Assert.Equal(new Notice("Download finished",
                                "https://youtu.be/e2\n1 touched, 0 already archived", false),
                     t.Update(new[] { Rec("e2", "done", touched: 1, skipped: 0, errors: 0), e1 }, 0));
    }

    [Fact]
    public void CancelledIsNamedInAQueue()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        // No run printed a summary, so there are no counts to give.
        Assert.Equal(new Notice("Queue finished: 1 done, 1 cancelled", "Nothing failed.", false),
                     t.Update(new[] { Rec("g2", "cancelled"), Rec("g1", "done") }, 0));
    }

    [Fact]
    public void RestoredHistoryIsNotAnnounced()
    {
        var restored = new[] { Rec("r1", "failed", "Interrupted") };
        var t = new NoticeTracker(restored);
        Assert.Null(t.Update(restored, 0));
    }

    [Fact]
    public void FailureReasons()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        var silent = Rec("h1", "failed");
        silent.ExitCode = 2;
        Assert.Equal("https://youtu.be/h1\nytdl exited with code 2.",
                     t.Update(new[] { silent }, 0)?.Body);

        var killed = Rec("h2", "failed", "  \n");
        killed.ExitCode = -1;
        Assert.Equal("https://youtu.be/h2\nIt printed no error before it stopped.",
                     t.Update(new[] { killed, silent }, 0)?.Body);
    }

    /// A record from a store written before options were saved has no URL;
    /// the command stands in for it.
    [Fact]
    public void TargetFallsBackToCommand()
    {
        var t = new NoticeTracker(Array.Empty<RunRecord>());
        var r = Rec("i1", "done");
        r.Opts = new RunOptions();
        Assert.Equal(new Notice("Download finished", "ytdl https://youtu.be/i1", false),
                     t.Update(new[] { r }, 0));
    }

    [Fact]
    public void Clip()
    {
        var c1 = NoticeTracker.Clip(new string('x', 200), 160);
        Assert.Equal(160, c1.EnumerateRunes().Count());
        Assert.EndsWith("x…", c1, StringComparison.Ordinal);

        // Code points, not UTF-16 units -- the count the C copy makes -- and a
        // character outside the BMP is never cut in half.
        var c2 = NoticeTracker.Clip(string.Concat(Enumerable.Repeat("é", 200)), 160);
        Assert.Equal(160, c2.EnumerateRunes().Count());
        var c3 = NoticeTracker.Clip(string.Concat(Enumerable.Repeat("😀", 200)), 160);
        Assert.Equal(160, c3.EnumerateRunes().Count());

        Assert.Equal("ERROR: a b c", NoticeTracker.Clip("  ERROR: a\nb\tc  ", 160));
        Assert.Equal("abc…", NoticeTracker.Clip("abc defgh", 5));
        Assert.Equal("short", NoticeTracker.Clip("short", 160));
    }
}

/* The Runner's side of it. Its own class, because it redirects the state
 * directory for its whole life. */
public sealed class RunnerSettledTests : IDisposable
{
    private readonly RedirectedHome _home;

    public RunnerSettledTests() => _home = new RedirectedHome("ytdl-win-settled");
    public void Dispose() => _home.Dispose();

    /* Settled must never report nothing left to run while a run is between
     * leaving the queue and being recorded as current. The worker takes the
     * item, drops the lock, and only then does RunOne set it current; in that
     * gap the queue is empty and nothing is running.
     *
     * Driven with runs that fail at once: the install root is an empty
     * directory, so each stops at "pwsh was not found" or "ytdl.ps1 does not
     * exist" having gone through the same take / Finish path a real run does,
     * spawning nothing. Polled from this thread as fast as it will go, like
     * the GTK copy.
     *
     * Probabilistic in the direction that matters: it cannot fail on correct
     * code, and it catches the bug some runs rather than every one. When it
     * was written, counting `_current is not null` instead of _inFlight failed
     * it 1 run in 5 here (the GTK copy: 3 in 5). */
    [Fact]
    public void SettledNeverReportsFinishedEarly()
    {
        const int n = 40;
        var runner = new Runner();
        var first = runner.Settled(0)!;
        Assert.Equal(0, first.Remaining);
        Assert.Empty(first.History);
        // Nothing changed since: nothing to report.
        Assert.Null(runner.Settled(first.Version));

        runner.SetPaused(true);
        for (var i = 0; i < n; i++)
            runner.Enqueue(new RunOptions { Url = "https://youtu.be/abcdefghijk" });
        // Paused with runs waiting is not finished.
        Assert.Equal(n, runner.Settled(0)!.Remaining);

        runner.Start();
        runner.SetPaused(false);

        var premature = 0;
        var finished = false;
        var clock = Stopwatch.StartNew();
        while (!finished && clock.Elapsed < TimeSpan.FromSeconds(20))
        {
            var s = runner.Settled(0)!;
            if (s.Remaining == 0)
            {
                if (s.History.Count < n) premature++;
                else finished = true;
            }
        }
        runner.Stop();

        Assert.True(finished);
        Assert.Equal(0, premature);
        var last = runner.Settled(0)!;
        Assert.Equal(n, last.History.Count);
        Assert.All(last.History, r => Assert.Equal("failed", r.State));
    }
}

File: macos-swiftui/Tests/NoticeTests.swift
/* What the queue announces, and when -- Core/Notices.swift's rules as
 * fixtures.
 *
 * THE SAME VALUES are asserted by linux-gtk/tests/test_notify.c and
 * windows-winui/YtdlWin.Tests/NoticeTests.cs. A title or a rule changed here
 * and not there is three apps announcing the same queue three different
 * ways, so the strings are spelled out in full rather than matched loosely.
 *
 * The last test drives the real Runner, to pin the one thing the tracker
 * cannot check for itself: that `remaining` never reads zero while a run is
 * about to start.
 */

import XCTest
@testable import YtdlMac

final class NoticeTests: XCTestCase {
    private func rec(_ id: String, _ state: String, _ lastLine: String = "",
                     touched: Int64 = -1, skipped: Int64 = -1, errors: Int64 = -1) -> RunRecord {
        var r = RunRecord()
        r.id = id
        r.state = state
        r.opts.url = "https://youtu.be/\(id)"
        r.command = "ytdl \(r.opts.url)"
        r.lastLine = lastLine
        r.videosTouched = touched
        r.archiveSkipped = skipped
        r.errors = errors
        r.warnings = 0
        r.exitCode = state == "failed" ? 1 : 0
        return r
    }

    func testSingleDone() {
        let t = NoticeTracker(history: [])
        let n = t.update(history: [rec("aaaaaaaaaaa", "done", touched: 3, skipped: 12, errors: 0)],
                         remaining: 0)
        XCTAssertEqual(n, Notice(title: "Download finished",
                                 body: "https://youtu.be/aaaaaaaaaaa\n3 touched, 12 already archived",
                                 failure: false))
    }

    func testSingleFailed() {
        let t = NoticeTracker(history: [])
        let n = t.update(history: [rec("bbbbbbbbbbb", "failed",
                                       "ERROR: [youtube] bbbbbbbbbbb: Private video")],
                         remaining: 0)
        XCTAssertEqual(n, Notice(title: "Download failed",
                                 body: "https://youtu.be/bbbbbbbbbbb\n"
                                     + "ERROR: [youtube] bbbbbbbbbbb: Private video",
                                 failure: true))
    }

    /// Three runs, one at a time, the middle one failing: silence, the failure
    /// at once, then one summary that repeats it.
    func testQueueOfThree() {
        let t = NoticeTracker(history: [])
        let c1 = rec("c1", "done", touched: 1, skipped: 0, errors: 0)
        let c2 = rec("c2", "failed", "ERROR: HTTP Error 403: Forbidden",
                     touched: 0, skipped: 0, errors: 1)
        let c3 = rec("c3", "done", touched: 2, skipped: 5, errors: 0)

        XCTAssertNil(t.update(history: [c1], remaining: 2))
        XCTAssertEqual(t.update(history: [c2, c1], remaining: 1),
                       Notice(title: "A download failed",
                              body: "https://youtu.be/c2\nERROR: HTTP Error 403: Forbidden\n"
                                  + "The queue goes on: 1 still to run.",
                              failure: true))
        // Nothing new: nothing to say, and the failure is not announced twice.
        XCTAssertNil(t.update(history: [c2, c1], remaining: 1))
        XCTAssertEqual(t.update(history: [c3, c2, c1], remaining: 0),
                       Notice(title: "Queue finished: 2 done, 1 failed",
                              body: "3 touched, 5 already archived, 1 error\n"
                                  + "Last failure: https://youtu.be/c2 — "
                                  + "ERROR: HTTP Error 403: Forbidden",
                              failure: true))
    }

    func testSecondFailureCounts() {
        let t = NoticeTracker(history: [])
        let f1 = rec("f1", "failed", "ERROR: one")
        let f2 = rec("f2", "failed", "ERROR: two")
        XCTAssertEqual(t.update(history: [f1], remaining: 3)?.title, "A download failed")
        XCTAssertEqual(t.update(history: [f2, f1], remaining: 2),
                       Notice(title: "2 downloads failed so far",
                              body: "https://youtu.be/f2\nERROR: two\n"
                                  + "The queue goes on: 2 still to run.",
                              failure: true))
    }

    /// A run finishing and the queue going idle in one drain tick: the summary
    /// only, never a failure notice followed by a summary.
    func testCoalescedTick() {
        let t = NoticeTracker(history: [])
        let n = t.update(history: [rec("d2", "failed", "ERROR: gone"),
                                   rec("d1", "done", touched: 4, skipped: 0, errors: 0)],
                         remaining: 0)
        XCTAssertEqual(n, Notice(title: "Queue finished: 1 done, 1 failed",
                                 body: "4 touched, 0 already archived\n"
                                     + "Last failure: https://youtu.be/d2 — ERROR: gone",
                                 failure: true))
    }

    func testCancelledOnlyIsSilent() {
        let t = NoticeTracker(history: [])
        let e1 = rec("e1", "cancelled")
        XCTAssertNil(t.update(history: [e1], remaining: 0))
        // ...and the queue it ended is closed: the next one does not count it.
        XCTAssertEqual(t.update(history: [rec("e2", "done", touched: 1, skipped: 0, errors: 0), e1],
                                remaining: 0),
                       Notice(title: "Download finished",
                              body: "https://youtu.be/e2\n1 touched, 0 already archived",
                              failure: false))
    }

    func testCancelledIsNamedInAQueue() {
        let t = NoticeTracker(history: [])
        // No run printed a summary, so there are no counts to give.
        XCTAssertEqual(t.update(history: [rec("g2", "cancelled"), rec("g1", "done")],
                                remaining: 0),
                       Notice(title: "Queue finished: 1 done, 1 cancelled",
                              body: "Nothing failed.", failure: false))
    }

    func testRestoredHistoryIsNotAnnounced() {
        let restored = [rec("r1", "failed", "Interrupted")]
        let t = NoticeTracker(history: restored)
        XCTAssertNil(t.update(history: restored, remaining: 0))
    }

    func testFailureReasons() {
        let t = NoticeTracker(history: [])
        var silent = rec("h1", "failed")
        silent.exitCode = 2
        XCTAssertEqual(t.update(history: [silent], remaining: 0)?.body,
                       "https://youtu.be/h1\nytdl exited with code 2.")

        var killed = rec("h2", "failed", "  \n")
        killed.exitCode = -1
        XCTAssertEqual(t.update(history: [killed, silent], remaining: 0)?.body,
                       "https://youtu.be/h2\nIt printed no error before it stopped.")
    }

    /// A record from a store written before options were saved has no URL;
    /// the command stands in for it.
    func testTargetFallsBackToCommand() {
        let t = NoticeTracker(history: [])
        var r = rec("i1", "done")
        r.opts = RunOptions()
        XCTAssertEqual(t.update(history: [r], remaining: 0),
                       Notice(title: "Download finished", body: "ytdl https://youtu.be/i1",
                              failure: false))
    }

    func testClip() {
        let c1 = NoticeTracker.clip(String(repeating: "x", count: 200), max: 160)
        XCTAssertEqual(c1.unicodeScalars.count, 160)
        XCTAssertTrue(c1.hasSuffix("x…"))

        // Characters, not bytes -- and the same count the C copy makes.
        let c2 = NoticeTracker.clip(String(repeating: "é", count: 200), max: 160)
        XCTAssertEqual(c2.unicodeScalars.count, 160)

        XCTAssertEqual(NoticeTracker.clip("  ERROR: a\nb\tc  ", max: 160), "ERROR: a b c")
        XCTAssertEqual(NoticeTracker.clip("abc defgh", max: 5), "abc…")
        XCTAssertEqual(NoticeTracker.clip("short", max: 160), "short")
    }

    // MARK: - The Runner's side of it

    /* Nothing-left-to-run must never be reported while a run is between
     * leaving the queue and being recorded as current. The worker takes the
     * item, drops the lock, and only then does runOne set it current; in that
     * gap the queue is empty and nothing is running.
     *
     * Driven with runs that fail at once -- HOME and the install root point at
     * an empty directory, so each stops at "pwsh not found" or "ytdl.ps1 does
     * not exist" having gone through the same take / finish path a real run
     * does, spawning nothing -- and polled through settled() as fast as this
     * thread will go, like the GTK and C# copies. (An earlier draft watched
     * onSettled instead, and passed 8 runs of 8 with the bug put back: a 50ms
     * drain almost never lands in a gap that wide.) Probabilistic in the
     * direction that matters: it cannot fail on correct code. With
     * `pendingCurrent != nil` counted instead of pendingInFlight it failed 3
     * runs in 6, each time seeing the gap over a thousand times. */
    func testSettledNeverReportsFinishedEarly() throws {
        let dir = try FixtureSupport.makeTempDir(prefix: "ytdl-macos-notice")
        /* Through the override rather than setenv, like every other test
         * here that must not touch ~/Library: the Runner reads and writes its
         * queue under Paths.stateDir(). Cleared only after stop(), because
         * the worker reads it too. */
        Paths.environmentOverride = ["HOME": dir, "YTDLP_INSTALL_ROOT": dir]
        defer {
            Paths.environmentOverride = nil
            try? FileManager.default.removeItem(atPath: dir)
        }

        let runner = Runner()
        XCTAssertEqual(runner.settled().remaining, 0)

        let n = 40
        runner.setPaused(true)
        var o = RunOptions()
        o.url = "https://youtu.be/abcdefghijk"
        for _ in 0..<n { _ = try runner.enqueue(o) }
        // Paused with runs waiting is not finished.
        XCTAssertEqual(runner.settled().remaining, n)

        runner.start()
        runner.setPaused(false)

        var premature = 0
        var finished = false
        let deadline = Date().addingTimeInterval(20)
        while !finished && Date() < deadline {
            let s = runner.settled()
            if s.remaining == 0 {
                if s.history.count < n { premature += 1 } else { finished = true }
            }
        }

        /* The drain publishes the same pair, and hands it to onSettled. */
        var delivered: (Int, Int)?
        runner.onSettled = { history, remaining in delivered = (history.count, remaining) }
        RunLoop.main.run(until: Date().addingTimeInterval(0.2))
        runner.stop()

        XCTAssertTrue(finished)
        XCTAssertEqual(premature, 0)
        XCTAssertEqual(runner.remaining, 0)
        XCTAssertEqual(runner.history.count, n)
        XCTAssertTrue(runner.history.allSatisfy { $0.state == "failed" })
        XCTAssertEqual(delivered?.0, n)
        XCTAssertEqual(delivered?.1, 0)
    }
}
